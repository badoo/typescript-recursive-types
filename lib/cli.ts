import {
    ModuleKind,
    ScriptTarget,
} from 'typescript';

import { generateDocumentationFromFiles } from './index';

console.log(JSON.stringify(generateDocumentationFromFiles(process.argv.slice(2), {
    target: ScriptTarget.ES5,
    module: ModuleKind.CommonJS,
}), null, 2));
