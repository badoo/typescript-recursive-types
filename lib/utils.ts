import {
    Declaration,
    getCombinedModifierFlags,
    LiteralType,
    ModifierFlags,
    Node,
    Symbol,
    SymbolFlags,
    SyntaxKind,
    Type,
    TypeChecker,
    TypeFlags,
    UnionOrIntersectionType,
} from 'typescript';

import { ObservedType } from './types';

/** True if this is visible outside this file, false otherwise */
export function isNodeExported(node: Node): boolean {
    return (
        (getCombinedModifierFlags(node as Declaration) & ModifierFlags.Export) !== 0 ||
        (!!node.parent && node.parent.kind === SyntaxKind.SourceFile)
    );
}

export function isBoolean(type: Type): type is UnionOrIntersectionType {
    return !!(type.getFlags() & TypeFlags.Boolean); // Boolean
}

export function isArray(type: ObservedType, checker: TypeChecker): type is ObservedType {
    const name = checker.typeToString(type);
    return name === '[]' || (type.symbol && type.symbol.getName() === 'Array');
}

export function isFunction(type: Type) {
    return type.getCallSignatures().length > 0;
}

export function isRequired(symbol: Symbol) {
    return !(symbol.getFlags() & SymbolFlags.Optional);
}

export function isBooleanLiteral(type: Type): type is LiteralType {
    return !!(type.getFlags() & TypeFlags.BooleanLiteral);
}
