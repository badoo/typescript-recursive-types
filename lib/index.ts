import {
    CompilerOptions,
    createProgram,
    displayPartsToString,
    forEachChild,
    isClassDeclaration,
    isModuleDeclaration,
    Node,
    Signature,
    Symbol,
    Type,
    TypeChecker,
} from 'typescript';

import {
    DocEntry,
    DocEntryContext,
    DocEntryType,
    DocGenParams,
    ObservedSymbol,
    ObservedType,
} from './types';

import { isNodeExported, isBoolean, isArray, isFunction, isRequired, isBooleanLiteral } from './utils';

const __DEBUG__ = false;

export {
    generateDocumentationFromNode,
    generateDocumentationFromFiles,
    serializeType,
    serializeSymbol,
    serializeClass,
    serializeSignature,
};

export default {
    generateDocumentationFromNode,
    generateDocumentationFromFiles,
    serializeType,
    serializeSymbol,
    serializeClass,
    serializeSignature,
};

/** Generate description for all classes in a set of .ts files */
function generateDocumentationFromNode(
    checker: TypeChecker,
    node: Node,
    docGenParams?: DocGenParams
): DocEntry[] {
    return visit(checker, node, docGenParams);
}

function generateDocumentationFromFiles(
    fileNames: string[],
    options: CompilerOptions,
    docGenParams?: DocGenParams
): DocEntry[] {
    // Build a program using the set of root file names in fileNames
    let program = createProgram(fileNames, options);

    // Get the checker, we will use it to find more about classes
    let checker = program.getTypeChecker();

    let output: DocEntry[] = [];

    // Visit every sourceFile in the program
    for (const sourceFile of program.getSourceFiles()) {
        if (!sourceFile.isDeclarationFile) {
            // Walk the tree to search for classes
            forEachChild(sourceFile, (node: Node) => {
                output = output.concat(generateDocumentationFromNode(checker, node, docGenParams));
            });
        }
    }

    return output;
}

/** visit nodes finding exported classes */
function visit(checker: TypeChecker, node: Node, docGenParams?: DocGenParams) {
    const output: DocEntry[] = [];

    // Only consider exported nodes
    if (!isNodeExported(node)) {
        return output;
    }

    if (isClassDeclaration(node) && node.name) {
        // This is a top level class, get its symbol
        const symbol = checker.getSymbolAtLocation(node.name);
        const docContext: DocEntryContext = {
            serialisedTypes: [],
            checker,
            maxDepth: docGenParams && docGenParams.maxDepth ? docGenParams.maxDepth : 5,
            maxProps: docGenParams && docGenParams.maxProps ? docGenParams.maxProps : 30,
        };

        if (symbol) {
            output.push(serializeClass(docContext, symbol));
        }
        // No need to walk any further, class expressions/inner declarations
        // cannot be exported
    } else if (isModuleDeclaration(node)) {
        // This is a namespace, visit its children
        forEachChild(node, visit.bind(null, checker));
    }

    return output;
}

/** Serialize a symbol into a json object */
function serializeType(docContext: DocEntryContext, type: Type): DocEntryType {
    const checker = docContext.checker;
    const raw = checker.typeToString(type);
    const symbol = type.symbol;
    const name = symbol ? symbol.getName() : raw;
    const description = symbol
        ? displayPartsToString(symbol.getDocumentationComment(docContext.checker))
        : '';

    // We've already serialised the type in this stack
    // This is needed to prevent infinite loop on recursive or circular depencency types
    const stopNestingTypes =
        docContext.serialisedTypes.includes(type) ||
        docContext.serialisedTypes.length > docContext.maxDepth;

    if (__DEBUG__) {
        console.log('Type', raw);
    }

    // We don't want to mutate the context as it's shared
    const newDocContext = {
        ...docContext,
        serialisedTypes: [...docContext.serialisedTypes, type],
    };

    // In TS booleans behave like enums so we need to filter them out earlier
    if (isBoolean(type)) {
        return {
            name,
            raw,
            type: 'boolean',
            description,
            value: stopNestingTypes ? [] : type.types.map(serializeType.bind(null, newDocContext)),
        };
    }

    if (isArray(type, checker)) {
        return {
            name,
            raw,
            type: 'array',
            description,
            value: stopNestingTypes
                ? []
                : type.typeArguments
                ? type.typeArguments.map(serializeType.bind(null, newDocContext))
                : 'none',
        };
    }

    if (isFunction(type)) {
        return {
            name,
            raw,
            type: 'function',
            description,
            value: raw,
            callSignature: stopNestingTypes
                ? []
                : type.getCallSignatures().map(serializeSignature.bind(null, newDocContext)),
        };
    }

    // Regular union or intersection enums
    if (type.isUnionOrIntersection()) {
        return {
            name,
            raw,
            type: 'enum',
            description,
            value: stopNestingTypes ? [] : type.types.map(serializeType.bind(null, newDocContext)),
        };
    }

    // 'as const' or other types like that @todo
    if ((type as ObservedType).typeArguments) {
        return {
            name,
            raw,
            type: 'enum',
            description,
            value: stopNestingTypes
                ? []
                : (type as ObservedType).typeArguments!.map(
                      serializeType.bind(null, newDocContext)
                  ),
        };
    }

    if (!symbol || type.isLiteral()) {
        let value: number | string = 'unknown';
        let detectedType = 'unknown';

        if (type.isStringLiteral()) {
            value = type.value;
            detectedType = 'string';
        } else if (type.isNumberLiteral()) {
            value = type.value;
            detectedType = 'number';
        } else if (isBooleanLiteral(type)) {
            value = raw;
            detectedType = 'boolean';
        } else if (typeof (type as ObservedType).intrinsicName === 'string') {
            value = (type as ObservedType).intrinsicName as string;
            detectedType = 'string';
        }

        if (raw === 'any' && value === 'any') {
            detectedType = 'any';
        }

        return {
            name,
            raw,
            type: detectedType,
            description,
            value,
        };
    }

    const properties = type.getProperties();

    if (!properties) {
        throw new Error(`Expected type to have some properties: ${raw}`);
    }

    const mappedProperties: DocEntryType[] = [];

    // Check if max number of props has exceeded
    if (!stopNestingTypes && properties.length <= newDocContext.maxProps) {
        properties.forEach(symbol => {
            mappedProperties.push(serializeSymbol(docContext, symbol));
        });
    }

    return {
        name,
        raw,
        type: 'shape',
        description,
        value: mappedProperties,
    };
}

/** Serialize a symbol into a json object */
function serializeSymbol(docContext: DocEntryContext, symbol: Symbol): DocEntryType {
    if (__DEBUG__) {
        console.log('Symbol', symbol.getName());
    }

    const type =
        (symbol as ObservedSymbol).type ||
        docContext.checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration);

    const description = displayPartsToString(symbol.getDocumentationComment(docContext.checker));

    return {
        ...serializeType(docContext, type),
        description,
        required: isRequired(symbol),
        name: symbol.getName(),
    };
}

/** Serialize a class symbol information */
function serializeClass(docContext: DocEntryContext, symbol: Symbol) {
    let details = {
        name: symbol.getName(),
        description: displayPartsToString(symbol.getDocumentationComment(docContext.checker)),
        type: serializeSymbol(docContext, symbol),
    } as DocEntry;

    // Get the construct signatures
    let constructorType = docContext.checker.getTypeOfSymbolAtLocation(
        symbol,
        symbol.valueDeclaration
    );

    details.constructors = constructorType
        .getConstructSignatures()
        .map(serializeSignature.bind(null, docContext));

    return details;
}

/** Serialize a signature (call or construct) */
function serializeSignature(docContext: DocEntryContext, signature: Signature): DocEntry {
    return {
        parameters: signature.parameters.map(serializeSymbol.bind(null, docContext)),
        returnType: docContext.checker.typeToString(signature.getReturnType()),
        description: displayPartsToString(signature.getDocumentationComment(docContext.checker)),
    };
}
