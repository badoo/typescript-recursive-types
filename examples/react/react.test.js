/* eslint-env jest */
import path from 'path';
import { ModuleKind, ScriptTarget } from 'typescript';
import { generateDocumentationFromFiles } from '../../dist/lib/index';

test('Extracts type information from file', () => {
    const testFile = path.resolve(__dirname, 'index.tsx');
    expect(
        generateDocumentationFromFiles([testFile], {
            target: ScriptTarget.ES5,
            module: ModuleKind.CommonJS,
        })
    ).toMatchSnapshot();
});
