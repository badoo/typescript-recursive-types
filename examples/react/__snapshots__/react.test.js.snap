// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extracts type information from file 1`] = `
Array [
  Object {
    "constructors": Array [
      Object {
        "description": "",
        "parameters": Array [
          Object {
            "description": "",
            "name": "props",
            "raw": "Readonly<Props>",
            "required": true,
            "type": "shape",
            "value": Array [
              Object {
                "description": "The type of button",
                "name": "type",
                "raw": "Types",
                "required": false,
                "type": "enum",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "\\"filled\\"",
                    "raw": "\\"filled\\"",
                    "type": "string",
                    "value": "filled",
                  },
                  Object {
                    "description": "",
                    "name": "\\"stroke\\"",
                    "raw": "\\"stroke\\"",
                    "type": "string",
                    "value": "stroke",
                  },
                  Object {
                    "description": "",
                    "name": "\\"transparent\\"",
                    "raw": "\\"transparent\\"",
                    "type": "string",
                    "value": "transparent",
                  },
                  Object {
                    "description": "",
                    "name": "\\"semitransparent\\"",
                    "raw": "\\"semitransparent\\"",
                    "type": "string",
                    "value": "semitransparent",
                  },
                  Object {
                    "description": "",
                    "name": "\\"monochrome\\"",
                    "raw": "\\"monochrome\\"",
                    "type": "string",
                    "value": "monochrome",
                  },
                  Object {
                    "description": "",
                    "name": "\\"link\\"",
                    "raw": "\\"link\\"",
                    "type": "string",
                    "value": "link",
                  },
                ],
              },
              Object {
                "description": "If the button should fit the smallest space available (\\"inline\\")",
                "name": "narrow",
                "raw": "boolean",
                "required": false,
                "type": "boolean",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "false",
                    "raw": "false",
                    "type": "boolean",
                    "value": "false",
                  },
                  Object {
                    "description": "",
                    "name": "true",
                    "raw": "true",
                    "type": "boolean",
                    "value": "true",
                  },
                ],
              },
              Object {
                "description": "Color applied to the button",
                "name": "color",
                "raw": "false | \\"primary\\" | \\"secondary\\" | \\"generic-red\\"",
                "required": false,
                "type": "enum",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "false",
                    "raw": "false",
                    "type": "boolean",
                    "value": "false",
                  },
                  Object {
                    "description": "",
                    "name": "\\"primary\\"",
                    "raw": "\\"primary\\"",
                    "type": "string",
                    "value": "primary",
                  },
                  Object {
                    "description": "",
                    "name": "\\"secondary\\"",
                    "raw": "\\"secondary\\"",
                    "type": "string",
                    "value": "secondary",
                  },
                  Object {
                    "description": "",
                    "name": "\\"generic-red\\"",
                    "raw": "\\"generic-red\\"",
                    "type": "string",
                    "value": "generic-red",
                  },
                ],
              },
              Object {
                "description": "Name of the icon to show inside the button",
                "name": "icon",
                "raw": "ReactNode",
                "required": true,
                "type": "enum",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "string",
                    "raw": "string",
                    "type": "string",
                    "value": "string",
                  },
                  Object {
                    "description": "",
                    "name": "number",
                    "raw": "number",
                    "type": "string",
                    "value": "number",
                  },
                  Object {
                    "description": "",
                    "name": "false",
                    "raw": "false",
                    "type": "boolean",
                    "value": "false",
                  },
                  Object {
                    "description": "",
                    "name": "true",
                    "raw": "true",
                    "type": "boolean",
                    "value": "true",
                  },
                  Object {
                    "description": "",
                    "name": "__type",
                    "raw": "{}",
                    "type": "shape",
                    "value": Array [],
                  },
                  Object {
                    "description": "",
                    "name": "ReactElement",
                    "raw": "ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>",
                    "type": "enum",
                    "value": Array [
                      Object {
                        "description": "",
                        "name": "any",
                        "raw": "any",
                        "type": "any",
                        "value": "any",
                      },
                      Object {
                        "description": "",
                        "name": "string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)",
                        "raw": "string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)",
                        "type": "enum",
                        "value": Array [
                          Object {
                            "description": "",
                            "name": "string",
                            "raw": "string",
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "callSignature": Array [
                              Object {
                                "description": "",
                                "parameters": Array [
                                  Object {
                                    "description": "",
                                    "name": "props",
                                    "raw": "any",
                                    "required": true,
                                    "type": "any",
                                    "value": "any",
                                  },
                                ],
                                "returnType": "ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>",
                              },
                            ],
                            "description": "",
                            "name": "__type",
                            "raw": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                            "type": "function",
                            "value": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                          },
                          Object {
                            "description": "",
                            "name": "__type",
                            "raw": "new (props: any) => Component<any, any, any>",
                            "type": "shape",
                            "value": Array [],
                          },
                        ],
                      },
                    ],
                  },
                  Object {
                    "description": "",
                    "name": "ReactNodeArray",
                    "raw": "ReactNodeArray",
                    "type": "shape",
                    "value": Array [],
                  },
                  Object {
                    "description": "",
                    "name": "ReactPortal",
                    "raw": "ReactPortal",
                    "type": "shape",
                    "value": Array [
                      Object {
                        "description": "",
                        "name": "key",
                        "raw": "ReactText",
                        "required": true,
                        "type": "enum",
                        "value": Array [
                          Object {
                            "description": "",
                            "name": "string",
                            "raw": "string",
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "description": "",
                            "name": "number",
                            "raw": "number",
                            "type": "string",
                            "value": "number",
                          },
                        ],
                      },
                      Object {
                        "description": "",
                        "name": "children",
                        "raw": "ReactNode",
                        "required": true,
                        "type": "enum",
                        "value": Array [],
                      },
                      Object {
                        "description": "",
                        "name": "type",
                        "raw": "string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)",
                        "required": true,
                        "type": "enum",
                        "value": Array [
                          Object {
                            "description": "",
                            "name": "string",
                            "raw": "string",
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "callSignature": Array [
                              Object {
                                "description": "",
                                "parameters": Array [
                                  Object {
                                    "description": "",
                                    "name": "props",
                                    "raw": "any",
                                    "required": true,
                                    "type": "any",
                                    "value": "any",
                                  },
                                ],
                                "returnType": "ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>",
                              },
                            ],
                            "description": "",
                            "name": "__type",
                            "raw": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                            "type": "function",
                            "value": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                          },
                          Object {
                            "description": "",
                            "name": "__type",
                            "raw": "new (props: any) => Component<any, any, any>",
                            "type": "shape",
                            "value": Array [],
                          },
                        ],
                      },
                      Object {
                        "description": "",
                        "name": "props",
                        "raw": "any",
                        "required": true,
                        "type": "any",
                        "value": "any",
                      },
                    ],
                  },
                ],
              },
              Object {
                "description": "Text to show inside the button",
                "name": "text",
                "raw": "string",
                "required": true,
                "type": "string",
                "value": "string",
              },
              Object {
                "description": "AB TEST",
                "name": "test",
                "raw": "ABTest",
                "required": false,
                "type": "shape",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "list",
                    "raw": "any[]",
                    "required": true,
                    "type": "array",
                    "value": Array [
                      Object {
                        "description": "",
                        "name": "any",
                        "raw": "any",
                        "type": "any",
                        "value": "any",
                      },
                    ],
                  },
                  Object {
                    "callSignature": Array [
                      Object {
                        "description": "",
                        "parameters": Array [
                          Object {
                            "description": "",
                            "name": "arg0",
                            "raw": "string",
                            "required": true,
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "description": "",
                            "name": "args",
                            "raw": "any[]",
                            "required": true,
                            "type": "array",
                            "value": Array [
                              Object {
                                "description": "",
                                "name": "any",
                                "raw": "any",
                                "type": "any",
                                "value": "any",
                              },
                            ],
                          },
                        ],
                        "returnType": "void",
                      },
                    ],
                    "description": "",
                    "name": "reportHit",
                    "raw": "(arg0: string, ...args: any[]) => void",
                    "required": true,
                    "type": "function",
                    "value": "(arg0: string, ...args: any[]) => void",
                  },
                ],
              },
              Object {
                "callSignature": Array [
                  Object {
                    "description": "",
                    "parameters": Array [
                      Object {
                        "description": "",
                        "name": "event",
                        "raw": "MouseEvent",
                        "required": true,
                        "type": "shape",
                        "value": Array [],
                      },
                    ],
                    "returnType": "boolean",
                  },
                ],
                "description": "\\"onClick\\" handler attached to the element",
                "name": "onClick",
                "raw": "(event: MouseEvent) => boolean",
                "required": true,
                "type": "function",
                "value": "(event: MouseEvent) => boolean",
              },
              Object {
                "callSignature": Array [
                  Object {
                    "description": "",
                    "parameters": Array [],
                    "returnType": "void",
                  },
                ],
                "description": "\\"onTouchStart\\" handler attached to the element",
                "name": "onTouchStart",
                "raw": "() => void",
                "required": true,
                "type": "function",
                "value": "() => void",
              },
            ],
          },
        ],
        "returnType": "Button",
      },
      Object {
        "description": "",
        "parameters": Array [
          Object {
            "description": "",
            "name": "props",
            "raw": "Props",
            "required": true,
            "type": "shape",
            "value": Array [
              Object {
                "description": "The type of button",
                "name": "type",
                "raw": "Types",
                "required": false,
                "type": "enum",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "\\"filled\\"",
                    "raw": "\\"filled\\"",
                    "type": "string",
                    "value": "filled",
                  },
                  Object {
                    "description": "",
                    "name": "\\"stroke\\"",
                    "raw": "\\"stroke\\"",
                    "type": "string",
                    "value": "stroke",
                  },
                  Object {
                    "description": "",
                    "name": "\\"transparent\\"",
                    "raw": "\\"transparent\\"",
                    "type": "string",
                    "value": "transparent",
                  },
                  Object {
                    "description": "",
                    "name": "\\"semitransparent\\"",
                    "raw": "\\"semitransparent\\"",
                    "type": "string",
                    "value": "semitransparent",
                  },
                  Object {
                    "description": "",
                    "name": "\\"monochrome\\"",
                    "raw": "\\"monochrome\\"",
                    "type": "string",
                    "value": "monochrome",
                  },
                  Object {
                    "description": "",
                    "name": "\\"link\\"",
                    "raw": "\\"link\\"",
                    "type": "string",
                    "value": "link",
                  },
                ],
              },
              Object {
                "description": "If the button should fit the smallest space available (\\"inline\\")",
                "name": "narrow",
                "raw": "boolean",
                "required": false,
                "type": "boolean",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "false",
                    "raw": "false",
                    "type": "boolean",
                    "value": "false",
                  },
                  Object {
                    "description": "",
                    "name": "true",
                    "raw": "true",
                    "type": "boolean",
                    "value": "true",
                  },
                ],
              },
              Object {
                "description": "Color applied to the button",
                "name": "color",
                "raw": "false | \\"primary\\" | \\"secondary\\" | \\"generic-red\\"",
                "required": false,
                "type": "enum",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "false",
                    "raw": "false",
                    "type": "boolean",
                    "value": "false",
                  },
                  Object {
                    "description": "",
                    "name": "\\"primary\\"",
                    "raw": "\\"primary\\"",
                    "type": "string",
                    "value": "primary",
                  },
                  Object {
                    "description": "",
                    "name": "\\"secondary\\"",
                    "raw": "\\"secondary\\"",
                    "type": "string",
                    "value": "secondary",
                  },
                  Object {
                    "description": "",
                    "name": "\\"generic-red\\"",
                    "raw": "\\"generic-red\\"",
                    "type": "string",
                    "value": "generic-red",
                  },
                ],
              },
              Object {
                "description": "Name of the icon to show inside the button",
                "name": "icon",
                "raw": "ReactNode",
                "required": true,
                "type": "enum",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "string",
                    "raw": "string",
                    "type": "string",
                    "value": "string",
                  },
                  Object {
                    "description": "",
                    "name": "number",
                    "raw": "number",
                    "type": "string",
                    "value": "number",
                  },
                  Object {
                    "description": "",
                    "name": "false",
                    "raw": "false",
                    "type": "boolean",
                    "value": "false",
                  },
                  Object {
                    "description": "",
                    "name": "true",
                    "raw": "true",
                    "type": "boolean",
                    "value": "true",
                  },
                  Object {
                    "description": "",
                    "name": "__type",
                    "raw": "{}",
                    "type": "shape",
                    "value": Array [],
                  },
                  Object {
                    "description": "",
                    "name": "ReactElement",
                    "raw": "ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>",
                    "type": "enum",
                    "value": Array [
                      Object {
                        "description": "",
                        "name": "any",
                        "raw": "any",
                        "type": "any",
                        "value": "any",
                      },
                      Object {
                        "description": "",
                        "name": "string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)",
                        "raw": "string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)",
                        "type": "enum",
                        "value": Array [
                          Object {
                            "description": "",
                            "name": "string",
                            "raw": "string",
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "callSignature": Array [
                              Object {
                                "description": "",
                                "parameters": Array [
                                  Object {
                                    "description": "",
                                    "name": "props",
                                    "raw": "any",
                                    "required": true,
                                    "type": "any",
                                    "value": "any",
                                  },
                                ],
                                "returnType": "ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>",
                              },
                            ],
                            "description": "",
                            "name": "__type",
                            "raw": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                            "type": "function",
                            "value": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                          },
                          Object {
                            "description": "",
                            "name": "__type",
                            "raw": "new (props: any) => Component<any, any, any>",
                            "type": "shape",
                            "value": Array [],
                          },
                        ],
                      },
                    ],
                  },
                  Object {
                    "description": "",
                    "name": "ReactNodeArray",
                    "raw": "ReactNodeArray",
                    "type": "shape",
                    "value": Array [],
                  },
                  Object {
                    "description": "",
                    "name": "ReactPortal",
                    "raw": "ReactPortal",
                    "type": "shape",
                    "value": Array [
                      Object {
                        "description": "",
                        "name": "key",
                        "raw": "ReactText",
                        "required": true,
                        "type": "enum",
                        "value": Array [
                          Object {
                            "description": "",
                            "name": "string",
                            "raw": "string",
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "description": "",
                            "name": "number",
                            "raw": "number",
                            "type": "string",
                            "value": "number",
                          },
                        ],
                      },
                      Object {
                        "description": "",
                        "name": "children",
                        "raw": "ReactNode",
                        "required": true,
                        "type": "enum",
                        "value": Array [],
                      },
                      Object {
                        "description": "",
                        "name": "type",
                        "raw": "string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)",
                        "required": true,
                        "type": "enum",
                        "value": Array [
                          Object {
                            "description": "",
                            "name": "string",
                            "raw": "string",
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "callSignature": Array [
                              Object {
                                "description": "",
                                "parameters": Array [
                                  Object {
                                    "description": "",
                                    "name": "props",
                                    "raw": "any",
                                    "required": true,
                                    "type": "any",
                                    "value": "any",
                                  },
                                ],
                                "returnType": "ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>",
                              },
                            ],
                            "description": "",
                            "name": "__type",
                            "raw": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                            "type": "function",
                            "value": "(props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>",
                          },
                          Object {
                            "description": "",
                            "name": "__type",
                            "raw": "new (props: any) => Component<any, any, any>",
                            "type": "shape",
                            "value": Array [],
                          },
                        ],
                      },
                      Object {
                        "description": "",
                        "name": "props",
                        "raw": "any",
                        "required": true,
                        "type": "any",
                        "value": "any",
                      },
                    ],
                  },
                ],
              },
              Object {
                "description": "Text to show inside the button",
                "name": "text",
                "raw": "string",
                "required": true,
                "type": "string",
                "value": "string",
              },
              Object {
                "description": "AB TEST",
                "name": "test",
                "raw": "ABTest",
                "required": false,
                "type": "shape",
                "value": Array [
                  Object {
                    "description": "",
                    "name": "list",
                    "raw": "any[]",
                    "required": true,
                    "type": "array",
                    "value": Array [
                      Object {
                        "description": "",
                        "name": "any",
                        "raw": "any",
                        "type": "any",
                        "value": "any",
                      },
                    ],
                  },
                  Object {
                    "callSignature": Array [
                      Object {
                        "description": "",
                        "parameters": Array [
                          Object {
                            "description": "",
                            "name": "arg0",
                            "raw": "string",
                            "required": true,
                            "type": "string",
                            "value": "string",
                          },
                          Object {
                            "description": "",
                            "name": "args",
                            "raw": "any[]",
                            "required": true,
                            "type": "array",
                            "value": Array [
                              Object {
                                "description": "",
                                "name": "any",
                                "raw": "any",
                                "type": "any",
                                "value": "any",
                              },
                            ],
                          },
                        ],
                        "returnType": "void",
                      },
                    ],
                    "description": "",
                    "name": "reportHit",
                    "raw": "(arg0: string, ...args: any[]) => void",
                    "required": true,
                    "type": "function",
                    "value": "(arg0: string, ...args: any[]) => void",
                  },
                ],
              },
              Object {
                "callSignature": Array [
                  Object {
                    "description": "",
                    "parameters": Array [
                      Object {
                        "description": "",
                        "name": "event",
                        "raw": "MouseEvent",
                        "required": true,
                        "type": "shape",
                        "value": Array [],
                      },
                    ],
                    "returnType": "boolean",
                  },
                ],
                "description": "\\"onClick\\" handler attached to the element",
                "name": "onClick",
                "raw": "(event: MouseEvent) => boolean",
                "required": true,
                "type": "function",
                "value": "(event: MouseEvent) => boolean",
              },
              Object {
                "callSignature": Array [
                  Object {
                    "description": "",
                    "parameters": Array [],
                    "returnType": "void",
                  },
                ],
                "description": "\\"onTouchStart\\" handler attached to the element",
                "name": "onTouchStart",
                "raw": "() => void",
                "required": true,
                "type": "function",
                "value": "() => void",
              },
            ],
          },
          Object {
            "description": "",
            "name": "context",
            "raw": "any",
            "required": true,
            "type": "any",
            "value": "any",
          },
        ],
        "returnType": "Button",
      },
    ],
    "description": "",
    "name": "Button",
    "type": Object {
      "description": "",
      "name": "Button",
      "raw": "typeof Button",
      "required": true,
      "type": "shape",
      "value": Array [
        Object {
          "description": "",
          "name": "prototype",
          "raw": "any",
          "required": true,
          "type": "string",
          "value": "error",
        },
        Object {
          "description": "If set, \`this.context\` will be set at runtime to the current value of the given Context.

Usage:

\`\`\`ts
type MyContext = number
const Ctx = React.createContext<MyContext>(0)

class Foo extends React.Component {
   static contextType = Ctx
   context!: React.ContextType<typeof Ctx>
   render () {
     return <>My context's value: {this.context}</>;
   }
}
\`\`\`",
          "name": "contextType",
          "raw": "Context<any>",
          "required": false,
          "type": "enum",
          "value": Array [
            Object {
              "description": "",
              "name": "any",
              "raw": "any",
              "type": "any",
              "value": "any",
            },
          ],
        },
      ],
    },
  },
]
`;
